src/storage/mvcc/reader/reader.rs

```diff
@@ -580,15 +579,17 @@ impl<S: EngineSnapshot> MvccReader<S> {
                     break;
                 }
             }
-            let commit_ts = key.decode_ts()?;
             let user_key = key.clone().truncate_ts()?;
-            // To make sure we only check each unique key once and `filter(&key)` returns
+            let commit_ts = key.decode_ts()?;
+            // To make sure we only check each unique user key once and the filter returns
             // true.
-            if (cur_key.is_some() && cur_key.clone().unwrap() == user_key) || !filter(&key) {
+            let should_skip = (cur_key.is_some() && cur_key.clone().unwrap() == user_key)
+                || !filter(&user_key, commit_ts);
+            cur_key = Some(user_key.clone());
```

Comment:
Can you optimize some clones here?

`(cur_key.as_ref() == Some(&user_key))` can void the first clone. And we may not assign to `cur_key` if `cur_key` equals to `user_key`.